File: platform/app/server_store.py
"""JSON-file persistence for server specs."""

from __future__ import annotations

import json
from pathlib import Path

from app.config import SERVERS_DATA_DIR
from app.models import ServerSpec


class ServerStore:
    def __init__(self) -> None:
        self.base_dir = SERVERS_DATA_DIR

    def _spec_path(self, name: str) -> Path:
        return self.base_dir / name / "server.json"

    def save(self, spec: ServerSpec) -> None:
        path = self._spec_path(spec.name)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(spec.model_dump_json(indent=2))

    def load(self, name: str) -> ServerSpec | None:
        path = self._spec_path(name)
        if not path.exists():
            return None
        data = json.loads(path.read_text())
        return ServerSpec(**data)

    def delete(self, name: str) -> None:
        import shutil
        server_dir = self.base_dir / name
        if server_dir.exists():
            shutil.rmtree(server_dir)

    def list_all(self) -> list[ServerSpec]:
        results = []
        if not self.base_dir.exists():
            return results
        for server_dir in sorted(self.base_dir.iterdir()):
            spec_path = server_dir / "server.json"
            if spec_path.exists():
                data = json.loads(spec_path.read_text())
                results.append(ServerSpec(**data))
        return results

File: api-laravel/app/Services/Mcp/ServerStore.php
<?php

namespace App\Services\Mcp;

/**
 * JSON-file persistence for ServerSpecs - port of server_store.py.
 * Each server lives at {base_dir}/{name}/server.json
 */
class ServerStore
{
    public function __construct(private readonly string $baseDir) {}

    public function save(ServerSpec $spec): void
    {
        $dir = $this->serverDir($spec->name);
        if (! is_dir($dir)) {
            mkdir($dir, 0755, true);
        }
        $json = json_encode($spec->toArray(), JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);
        file_put_contents($this->specPath($spec->name), $json);
    }

    public function load(string $name): ?ServerSpec
    {
        $path = $this->specPath($name);
        if (! is_file($path)) {
            return null;
        }
        $raw = file_get_contents($path);
        if ($raw === false) {
            return null;
        }
        $data = json_decode($raw, true);
        if (! is_array($data)) {
            return null;
        }
        return ServerSpec::fromArray($data);
    }

    public function delete(string $name): void
    {
        $dir = $this->serverDir($name);
        if (is_dir($dir)) {
            $this->rrmdir($dir);
        }
    }

    /** @return ServerSpec[] */
    public function listAll(): array
    {
        if (! is_dir($this->baseDir)) {
            return [];
        }
        $out = [];
        $entries = scandir($this->baseDir) ?: [];
        sort($entries);
        foreach ($entries as $entry) {
            if ($entry === '.' || $entry === '..') {
                continue;
            }
            $spec = $this->load($entry);
            if ($spec) {
                $out[] = $spec;
            }
        }
        return $out;
    }

    public function serverDir(string $name): string
    {
        return $this->baseDir.'/'.$name;
    }

    private function specPath(string $name): string
    {
        return $this->serverDir($name).'/server.json';
    }

    private function rrmdir(string $dir): void
    {
        foreach (scandir($dir) as $entry) {
            if ($entry === '.' || $entry === '..') {
                continue;
            }
            $path = $dir.'/'.$entry;
            is_dir($path) ? $this->rrmdir($path) : @unlink($path);
        }
        @rmdir($dir);
    }
}
